### ISSUE TITLE
Can you explain linear part in RGN

### ISSUE THREAD

**Issue by OsamaGhandour** [conradry/pytorch-rgn#2]
you are declare linear_size in function as 20 and in the bottom of the code you send it as 60 can you explain why ? and if you can explain linear and softmax part and what happend after get lstm output that would be very helpful and thanks in advance

**Comment by conradry** [comment]
The linear size is used to create an "alphabet" of torsion angles. To begin, the alphabet is initialized by taking a random sample of angles from -pi to pi radians (the samples are not evenly spaced). Instead of directly regressing an angle, the RGN acts sort of like a classifier (hence the softmax). As far as I know, the linear_size of 20 isn't special, it's just the value that worked best.

To summarize, LSTM output passes through the Linear layer. The output of Linear layer + Softmax determines the most likely torsion angle from those available in the trainable alphabet of 20 angles. Note that the softmax output is never argmaxed; this allows for the prediction of torsion angles that are linear combinations of the angles in the alphabet.

**Comment by OsamaGhandour** [comment]
how did you say that alphabet is trainable ?
you initialize it, then where at the code alphabet are being updated ?
and i think it should be 60x3 because your are using a 20 length protein sequence so you have to get 180 angle at the end? and that happend when you multiply 60 output from (linear+softmax) by alphabet
 i hope if you could make alphabet part is more clear @conradry

**Comment by conradry** [comment]
Alphabet is a [Parameter](https://pytorch.org/docs/stable/generated/torch.nn.parameter.Parameter.html): ```self.alphabet = nn.Parameter(u.rsample(torch.Size([linear_units, 3])))```, it's updated with all the other model parameters during the backward pass. The length of protein sequences is not 20; the sequences can have an arbitrary length (see cell 7 in rgn.ipynb). Examples are:

```(0, torch.Size([622, 32, 41]), torch.Size([1866, 32, 3]))
(1, torch.Size([676, 32, 41]), torch.Size([2028, 32, 3]))
(2, torch.Size([695, 32, 41]), torch.Size([2085, 32, 3]))
(3, torch.Size([731, 32, 41]), torch.Size([2193, 32, 3]))
```

Sequence lengths are 622, 676, 695, 731, all batch sizes are 32, 41 is the channel dimension. Targets are rightfully 3x longer.

The length of the alphabet (20, 60, etc.) is just a hyperparameter to tune. I used a length of 20 as a test, but the paper calls for a length of 60.

**Comment by OsamaGhandour** [comment]
I got now that the alphabet are being updated while training but I'm have other questions.
How did you train on different lengths protein although LSTM **input size must be fixed.**
**What is batch_size?** i think that every single residue represent as 1 vector of 41 value ?!
it is still difficult for me to understand what happened overall in the model but what I'm already got now is:
input sequence --> LSTM Bidirectional 2 Layer of size 800 --> fully connected dense of size 60 --> SoftMax layer of size 60 --> Multiply Softmax output * alphabet --> 3 angles for each residue --> pred_coord.
but i think i got some details wrong after lstm_output so ..
I hope you can make **this model is more clear** for me.
Thanks in advance. @conradry

**Comment by conradry** [comment]
LSTM input does not need to be fixed. The only requirement is that the length of all sequences within a single batch have to be the same (maybe you're used to Tensorflow? PyTorch builds the computation graph dynamically). Different batches can have examples with different lengths.

Batch size is 32. I.e. 32 protein sequences are collected from a dataset. Each sequence is N residues long, and each residue is a 41-d vector. Then all of these 32 sequences are padded with 0s so that they are the same length as the longest sequence in that batch of 32. To make sure that the padded 0s don't affect the calculated hidden states, we use pad_packed_sequences (these lines in the code):

```python
packed = pack_padded_sequence(pad_seq[:, order], lengths[order], batch_first=False)  
lstm_out, _ = self.lstm(packed, (h0,c0))
unpacked, _ = pad_packed_sequence(lstm_out, batch_first=False, padding_value=0.0)
```

(There are resources written elsewhere about pack_padded_sequence). Were these lines of code the ones that were confusing? They make the model run more efficiently but they do not perform any calculations on the LSTM output. Aside from this, you're understanding of the flow of data through the model is correct.

**Comment by OsamaGhandour** [comment]
It is almost clear now but this is my last question in this part of code 
        
        softmax_out = F.softmax(self.linear(unpacked), dim=2)
        sine = torch.matmul(softmax_out, torch.sin(self.alphabet))
        cosine = torch.matmul(softmax_out, torch.cos(self.alphabet))
        out = torch.atan2(sine, cosine)
        
        #create as many copies of first 3 coords as there are samples in the batch
        broadcast = torch.ones((batch_sz, 3))
        pred_coords = torch.stack([self.A*broadcast, self.B*broadcast, self.C*broadcast])
        
        for ix, triplet in enumerate(out[1:]):
            pred_coords = geometric_unit(pred_coords, triplet, 
                                         self.avg_bond_angles, 
                                         self.avg_bond_lens)
        #pred_coords.register_hook(self.save_grad('pc'))
        
            
        #pdb.set_trace()
        return pred_coords


out shape would be 3x60 due to alphabet shape and linear length .. but if the sequence length was for example 720 i think out should be 3x720 .. i don't see how out shape and pred_coords are match the protein length??
**i hope if you could make this output part is more clear**
thanks in advance. @conradry

**Comment by conradry** [comment]
@Osos8128 sorry for the long delay.

The output of ```softmax_out``` is (L, B, 60) where L is the number of residues e.g. 720 and B is batch size e.g. 32. In ```torch.matmul``` you have (L, B, 60) x (60, 3) --> (L, B, 3) which is the size of ```out```.

```pred_coords``` is initialized as (3, B, 3). Then looping through ```out``` gives triplets of size (B, 3). Every iteration of the loop adds 3 new coordinates to ```pred_coords```. E.g. after the first iteration ```pred_coords``` is size (6, B, 3). If the number of residues is 720 then the final output size is (720 * 3, B, 3) = (2160, B, 3).

**Comment by OsamaGhandour** [comment]
@conradry thanks a lot for your help it is all clear now and my model is almost starts training. 
but i have a final questions, on Qurashi's paper i found that he test a Free modeling proteins and as i know input of FM is only the amino acids. how he did that although, the model takes an PSSM+AA as an input ?!

**Comment by OsamaGhandour** [comment]
@conradry please try to reply on this question early. your dRMSD function which shape expect ?
(seq_len,batch_size,3) or(batch_size,seq_len,3) ?? also the loss that return from your function are in m or nm ? because i think it should be converted to angstrom ?!

**Comment by conradry** [comment]
For the PSSM+AA input I think you'll need to use the ProteinNet dataset: https://github.com/aqlaboratory/proteinnet.

The dRMSD loss expect (seq_len, batch_size, 3) as input, then first few lines of the forward function permute the order to (batch_size, seq_len, 3).

The units in PDB files are Angstroms. In the ProteinNet dataset they are converted to nanometers (i.e. multiplied by 100). The units of the dRMSD function depend on the units of the coordinates you are using for the ground truth coordinates.

**Comment by OsamaGhandour** [comment]
I already using Sidechiannet data that is extend from proteinnet. So i have the input but i mean for a model trained using evo data how to make free modeling prediction ?
but i got the units part also i think you are wrong in those lines in dRMSD function.
    **idx = torch.tensor(mask[i][mask[i] != 0], dtype=torch.long)
    xdist_mat = pair_dist(torch.index_select(x[i], 0, idx))
    ydist_mat = pair_dist(torch.index_select(y[i], 0, idx))**
you want to avoid the missing residue crds but idx just contain ones not indices so index_select will just return atom[1] coordinates repeated.

also what is the point from this line while backprop

**torch.nn.utils.clip_grad_norm_(rgn.parameters(), max_norm=50)**

??
